alf/algorithms/merlin_algorithm.py

```diff
@@ -440,50 +443,94 @@ def calc_loss(self, training_info: TrainingInfo):
 
 
 @gin.configurable
-def create_merlin_algorithm(env,
-                            encoder_fc_layers=(3, ),
-                            latent_dim=3,
-                            lstm_size=(4, ),
-                            memory_size=20,
-                            learning_rate=1e-1,
-                            debug_summaries=True):
-    """Create a simple MerlinAlgorithm
+class ResnetEncodingNetwork(network.Network):
+    """Image encoding network
 
-    Args:
-        env (TFEnvironment): A TFEnvironment
-        encoder_fc_layers (list[int]): list of fc layers parameters for encoder
-        latent_dim (int): the dimension of the hidden representation of VAE.
-        lstm_size (list[int]): size of lstm layers for MBP and MBA
-        memory_size (int): number of memory slots
-        learning_rate (float): learning rate for training
+    This is not a generic network, it implements `ImageEncoder` described in
+    2.1.1 of "Unsupervised Predictive Memory in a Goal-Directed Agent"
     """
-    observation_spec = env.observation_spec()
-    action_spec = env.action_spec()
-
-    encoder = EncodingNetwork(
-        input_tensor_spec=observation_spec,
-        fc_layer_params=encoder_fc_layers,
-        activation_fn=None,
-        name="ObsEncoder")
-
-    decoder = DecodingAlgorithm(
-        decoder=EncodingNetwork(
-            input_tensor_spec=TensorSpec((latent_dim, ), dtype=tf.float32),
-            fc_layer_params=encoder_fc_layers,
-            activation_fn=None,
-            name="ObsDecoder"),
-        loss_weight=100.)
-
-    optimizer = tf.optimizers.Adam(learning_rate=learning_rate)
-
-    algorithm = MerlinAlgorithm(
-        action_spec=action_spec,
-        encoders=encoder,
-        decoders=decoder,
-        latent_dim=latent_dim,
-        lstm_size=lstm_size,
-        memory_size=memory_size,
-        optimizer=optimizer,
-        debug_summaries=debug_summaries)
-
-    return algorithm
+
+    def __init__(self, input_tensor_spec, name='ResnetEncodingNetwork'):
+        """Create a `ResnetEncodingNetwork` instance
+
+        Args:
+            input_tensor_spec (TensorSpec|nested TensorSpec): input observations spec.
+        """
+        super().__init__(input_tensor_spec, state_spec=(), name=name)
+
+        enc_layers = []
+
+        for i, stride in enumerate([2, 1, 2, 1, 2, 1]):
+            enc_layers.append(
+                resnet50_block.BottleneckBlock(
+                    kernel_size=(3, 3),
+                    filters=(64, 32, 64),
+                    strides=stride,
+                    name='block%d' % i))
```

Comment:
should `name` of this Network be included?

Reply:
it's under the scope of `ResnetEncodingNetwork`,  L453

Reply:
How is the scope used? Don’t you need to use ‘with self.name_scope’ in order to use the scope?

Reply:
`ResnetEncodingNetwork` is a subclass of `keras_network.Network` ,  these variables are created in `ResnetEncodingNetwork.__init__`, they are sub models  

<img src='https://user-images.githubusercontent.com/1267359/70211030-1509e480-176f-11ea-8ea4-760d46cbef6f.png' height=128>

Reply:
looked into source code,  variables (weights) of a Layer or Model are created at build phase (or first call),  and `Layer.call` or `Model.call` is  called under its scope, 

```python
with backend.name_scope(self._name_scope()):
  ...
  self.call(...)
  ...
```